tests/form_params_test.go

```diff
@@ -0,0 +1,269 @@
+
+package tests
+
+import (
+	"context"
+	"github.com/google/go-safeweb/testing/requesttesting"
+	"log"
+	"net/http"
+	"strconv"
+	"testing"
+)
+
+func TestSimpleFormParameters(t *testing.T) {
+	reqBody := "Veg=Potato&Fruit=Apple"
+	reqBodyLen := len(reqBody)
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: " + strconv.Itoa(reqBodyLen) + "\r\n" + "\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		req.ParseForm()
+		if req.Form["Veg"][0] != "Potato" && req.Form["Fruit"][0] != "Apple" {
+			t.Errorf("req.Form: want Potato and Apple, got %s and %s", req.Form["Veg"][0], req.Form["Fruit"][0])
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+}
+
+// Test whether passing a POST request with body and without Content-Length
+// yields a 400 Bad Request
+func TestFormParametersMissingContentLength(t *testing.T) {
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+}
+
+// Ensure passing a negative Content-Length or one that overflows results in a
+// 400 Bad Request
+func TestFormParametersBadContentLength(t *testing.T) {
+	reqBody := "veggie=potato"
+	postReq := []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: -1" + reqBody + "\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+
+	postReq = []byte("POST / HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" + "Content-Length: 10000000000000000000000000" + reqBody + "\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+}
```

Comment:
This should be two separate tests

Reply:
I'll make it table-driven since it performs the same checks on malformed Content-Lengths :)